src/components/List/List.js

```diff
@@ -1,21 +1,94 @@
-import React from 'react';
+/* eslint-disable jsx-a11y/no-noninteractive-tabindex */
+/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
+
+import React, { useState } from 'react';
 import PropTypes from 'prop-types';
 import bem from 'bem';
 import ListItem from './ListItem';
 import styles from './List.scss';
+import { LIST_NAVIGATION_DIRECTIONS, ENTER_KEY } from '../../constants';
 
 const { block, elem } = bem({
     name: 'List',
     classnames: styles,
     propsToMods: ['isDivided']
 });
 
+const isListItem = element => element && element.type !== ListItem && element.type !== 'li';
+const NAVIGATION_STEP_VALUES = {
+    [LIST_NAVIGATION_DIRECTIONS.UP]: -1,
+    [LIST_NAVIGATION_DIRECTIONS.DOWN]: 1
+};
+
 const List = React.forwardRef((props, ref) => {
-    const { children, isDivided, ...rest } = props;
+    const { children, isDivided, onNavigate, onSelect, ...rest } = props;
+    const [selectedIndex, setSelectedIndex] = useState(null);
+
+    const getNextSelectedIndex = keyCode => {
+        const stepValue = NAVIGATION_STEP_VALUES[keyCode];
+        const nextSelectedIndex = selectedIndex + stepValue;
+
+        // Return 0 index if nextSelectedIndex has negative value or selectedIndex hasn't been updated before
+        if (nextSelectedIndex < 0 || selectedIndex === null) {
+            return 0;
+        }
+
+        // Return last React.Children index if nextSelectedIndex is out of the right bound
+        if (nextSelectedIndex >= children.length) {
+            return children.length - 1;
+        }
+
+        // Return nextSelectedIndex without any changes for others cases
+        return nextSelectedIndex;
+    };
+
+    const handleKeyDown = e => {
+        // Update selectedIndex with arrow navigation and make onNavigate function callback
+        if (e.key === LIST_NAVIGATION_DIRECTIONS.UP || e.key === LIST_NAVIGATION_DIRECTIONS.DOWN) {
+            const nextSelectedIndex = getNextSelectedIndex(e.key);
+
+            if (selectedIndex !== nextSelectedIndex) {
+                e.preventDefault();
+                setSelectedIndex(getNextSelectedIndex(e.key));
+
+                if (onNavigate) {
+                    onNavigate(e.key);
```

Comment:
It might be useful to return the nextIndex as well. Then the consuming component doesn't need to recalculate it. Perhaps can return an object with `{ navKey: e.key, nextSelectedIndex }` or just simply as the first argument